ksm.go

```diff
@@ -0,0 +1,497 @@
+		return err
+	}
+
+	return nil
+}
+
+func (k *ksm) restore() error {
+	var err error
+
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	if err = k.restoreSysFS(); err != nil {
+		return err
+	}
+
+	if err := k.run.close(); err != nil {
+		return err
+	}
+
+	if err := k.sleepInterval.close(); err != nil {
+		return err
+	}
+
+	if err := k.pagesToScan.close(); err != nil {
+		return err
+	}
+
+	k.initialized = false
+	return nil
+}
+
+func (k *ksm) tune(s ksmSetting) error {
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	newPagesToScan, err := s.pagesToScan()
+	if err != nil {
+		return err
+	}
+
+	if err = k.pagesToScan.write(newPagesToScan); err != nil {
+		return err
+	}
+
+	if err = k.sleepInterval.write(fmt.Sprintf("%v", s.scanIntervalMS)); err != nil {
+		return err
+	}
+
+	if err = k.run.write(ksmRun); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type ksmThrottleInterval struct {
+	interval time.Duration
+	nextKnob ksmMode
+}
+
+var ksmThrottleIntervals = map[ksmMode]ksmThrottleInterval{
+	ksmAggressive: {
+		interval: 30 * time.Second, // 30s before throttling down from aggressive to standard.
+		nextKnob: ksmStandard,
+	},
+
+	ksmStandard: {
+		interval: 120 * time.Second, // 120s before throttling down from standard to slow.
+		nextKnob: ksmSlow,
+	},
+
+	ksmSlow: {
+		interval: 0,          // We stop here unless a new container shows up
+		nextKnob: ksmInitial, // And we go back to our initial settings
+	},
+
+	// We should never make it here
+	ksmInitial: {
+		interval: 0, // We stay here unless a new container shows up
+	},
+}
+
+func (k *ksm) throttle() {
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		proxyLog.Error(errors.New("KSM is unavailable"))
+		return
+	}
+
+	k.currentKnob = ksmAggressive
+	k.throttling = true
+
+	go func() {
+		throttleTimer := time.NewTimer(ksmThrottleIntervals[k.currentKnob].interval)
+
+		for {
+			select {
+			case <-k.kickChannel:
+				// We got kicked, this means a new VM has been created.
+				// We will enter the aggressive setting until we throttle down.
+				_ = throttleTimer.Stop()
+				if err := k.tune(ksmSettings[ksmAggressive]); err != nil {
+					proxyLog.Error(err)
+				} else {
+					k.Lock()
+					k.currentKnob = ksmAggressive
+					k.Unlock()
+				}
+
+				_ = throttleTimer.Reset(ksmThrottleIntervals[k.currentKnob].interval)
+
+			case <-throttleTimer.C:
+				// Our throttling down timer kicked in.
+				// We will move down to the next knob and start the next time,
+				// if necessary.
+				var throttle = ksmThrottleIntervals[k.currentKnob]
+				if throttle.interval == 0 {
+					if throttle.nextKnob == ksmInitial {
+						k.Lock()
+						if err := k.restoreSysFS(); err != nil {
+							proxyLog.Error(err)
+						}
+						k.Unlock()
+					}
+					continue
+				}
+
+				nextKnob := throttle.nextKnob
+				if err := k.tune(ksmSettings[nextKnob]); err != nil {
```

Comment:
Just a note on something I noticed with KSM last week, and just verified. I don't think this will really affect the operation of `auto` mode here, but worth noting.
It looks like, from experimentation and reading the kernel code, that a write to KSMs `sleep_millisec` sysfs file does *not take effect until the previous setting triggers* - that is to say, if you had it set to 10s, and that had just fired, and then you set it to 100ms - that 100ms will not take effect until the pending 10s timer has fired and reset the internal kernel timer.
I suspect that is not intentional behaviour on the part of KSM, and would be better (more logical) if the timer got reset with every write to `sleep_millisec`. I'll contemplate a kernel patch :-)

Reply:
@grahamwhaley will turning KVM off, followed by on rearm the trigger with the right interval?

Reply:
you had to ask eh? ;-)
Yes, looks like it will. I believe you can do the disarm and new-interval in either order, so:
```
echo $new_timeout > sleepmillisecs
echo 0 > run
echo 1 > run
```
should do the right thing.
Further notes then:
- Just doing `echo 1 > run` when run==1 already will not trigger a new run afaict
- Writing 0 to run does not clear down any existing KSM merges (to do that you write 'bit 1) (value 2) into the run file.